### ISSUE TITLE
_npmUser.email field is not being updated

### ISSUE THREAD

**Issue by RyanCavanaugh** [npm/registry-issue-archive#56]
Root cause here: I want the small gravatar on this page https://www.npmjs.com/package/@types/jquery to show up correctly.

The NPM website is using the Gravatar for the initial email address I made the 'types' account with. You can see this in the NPM package details:

``` json
From https://registry.npmjs.org/@types%2Fjquery
{  
   "_id":"@types/jquery",
   "_rev":"16-ebc5d39f0c5937f95e2e40653fc7a4a9",
   "name":"@types/jquery",
   "description":"TypeScript definitions for jQuery 1.10.x / 2.0.x",
   "dist-tags":{  
      "latest":"2.0.32"
   },
   "versions":{  
      "1.10.8-alpha":{  
         /* snip snip */
         "_npmUser":{  
            "name":"types",
            "email":"ryan.cavanaugh@microsoft.com" <------- wrong
         },
```

However, this is not the email address we've been publishing with (this has been true for months), and the NPM website itself reports the correct email address (ts-npm-types@microsoft.com) https://www.npmjs.com/~types

This just seems like a bug; if not, how do I correctly publish the package such that the new email address is reported in the email field? We're using the NPM Client API if it matters (https://github.com/Microsoft/types-publisher/blob/master/src/lib/package-publisher.ts#L20)

**Comment by soldair** [comment]
would you email support@npmjs.com with the last 4 chars of the authentication token you are using to publish? We should be able to get to the bottom of it if we have that =)

**Comment by RyanCavanaugh** [comment]
The script authenticates using a password, not an auth token.

See https://github.com/Microsoft/types-publisher/blob/b285d22904af183946b6de6049f8a12d6b8eea2b/src/lib/npm-client.ts#L79 / https://github.com/Microsoft/types-publisher/blob/b285d22904af183946b6de6049f8a12d6b8eea2b/src/lib/npm-client.ts#L25

**Comment by soldair** [comment]
i would like the last 4 of this value 
https://github.com/Microsoft/types-publisher/blob/b285d22904af183946b6de6049f8a12d6b8eea2b/src/lib/npm-client.ts#L104

an aside. please save the token in settings instead of logging in each time. its probably unrelated but its how the other clients work and it'll make your commands one request faster.

**Comment by soldair** [comment]
Figured it out! Your custom publish process is putting the _npmUser with the ryan email in new versions. Our backend has no other way to get that email address because its not even tied to an npm account. 

Weirdly enough, at this time if the client sends an _npmUser value with a new version we allow it and use it to render who published the package on the website.

fixing this bug has been added to the queue, but to fix your gravatar now make sure your new publishes no not contain _npmUser so it can be populated by the registry backend with the authenticated user.

**Comment by ghost** [comment]
I put a few `console.logs` inside of our [`npm-client.ts`](https://github.com/Microsoft/types-publisher/blob/master/src/lib/npm-client.ts) and here's what I got:

The data we use to log in (`const params = ...`) is:

``` json
{
    "method": "PUT",
    "body": {
        "_id": "org.couchdb.user:types",
        "name": "types",
        "password": "swordfish",
        "type": "user",
        "roles": [],
        "date": "2016-09-29T13:42:43.664Z"
    }
}
```

The URL we are requesting is `https://registry.npmjs.org/-/user/org.couchdb.user:types`.

The metadata we publish (`const metadata = ...`) is:

``` json
{
    "readme": "# Installation\r\n> `npm install --save @types/amazon-product-api`\r\n\r\n# Summary\r\nThis package contains type definitions for amazon-product-api (https://github.com/t3chnoboy/amazon-product-api).\r\n\r\n# Details\r\nFiles were exported from https://www.github.com/DefinitelyTyped/DefinitelyTyped/tree/types-2.0/amazon-product-api\r\n\r\nAdditional Details\r\n * Last updated: Thu, 29 Sep 2016 13:39:55 GMT\r\n * File structure: ProperModule\r\n * Library Dependencies: none\r\n * Module Dependencies: none\r\n * Global values: none\r\n\r\n# Credits\r\nThese definitions were written by Matti Lehtinen <https://github.com/MattiLehtinen/>.\r\n",
    "name": "@types/amazon-product-api",
    "version": "0.0.31",
    "description": "TypeScript definitions for amazon-product-api",
    "license": "MIT",
    "author": "Matti Lehtinen <https://github.com/MattiLehtinen/>",
    "main": "",
    "repository": {
        "type": "git",
        "url": "https://www.github.com/DefinitelyTyped/DefinitelyTyped.git"
    },
    "scripts": {},
    "dependencies": {},
    "typings": "index.d.ts",
    "typesPublisherContentHash": "3a261d34f0544ea0b6249495755c0aee02cebd206c30ac06e36a1629738a4128"
}
```

It doesn't look like we're explicitly sending an `_npmUser` value, or if we are, `client.request` or `client.publish` is doing it for us. However, I couldn't find that string in the `npm-registry-client` source code, except in test cases.

**Comment by soldair** [comment]
its a little tough to debug with you because i dont have typescript to publish but.
if you could `console.log(JSON.stringify(metadata))` here
https://github.com/Microsoft/types-publisher/blob/master/src/lib/npm-client.ts#L36
or 
https://github.com/npm/npm-registry-client/blob/master/lib/publish.js#L35

it seems to me this bug would show up only where the document is in an odd state. like publishing from within node_modules  or where the metadata on disk somehow has _npmUser. i wouldnt expect it to be in the tests or code. it's just a value that gets passed through. the fix might be something as simple as `delete metadata._npmUser`

we are working on the fix internally and while it seemed like it would be super easy we found that testing to make sure we dont break a lot of folks is very complex because of some ancient logic in couchapp that only allows changes to specific object properties.

let me know if you figure anything else out. its a very interesting bug =)

**Comment by ghost** [comment]
I put a `console.log` in `node_modules/npm-registry-client/lib/publish.js` like you mentioned but it contains the same data as I posted above (under `const metadata = ...`).
We are publishing from files on disk in the `output` directory. The content in `output/amazon-product-api/package.json` is just the metadata without the `"readme"` field; no underscored names there.

**Comment by soldair** [comment]
and when you publish from the same environment you are testing in you still get the ryan@ email in _npmUser?

**Comment by ghost** [comment]
I just manually published `@types/ssh2` and it has the correct `_npmUser`.
Also, `_npmUser` is correct for the newest version of `@types/jquery` (2.0.32). Don't know what changed but :+1:.

However, the TypeScript avatar still isn't showing up on the [package](https://www.npmjs.com/package/@types/jquery) [pages](https://www.npmjs.com/package/@types/ssh2) (in front of "types published a week ago").

Also, the "maintainers" field still has the old email address. Our generated output does not contain that field, so it must be getting automatically added somehow.

**Comment by ghost** [comment]
By saving the token in settings, do you mean that we can safely continue using the same token forever?

**Comment by soldair** [comment]
yes. it will continue to work until someone sends it to a logout api call, we find it's been exposed in a package publish, or it's revoked manually.

the npm cli save these tokens in the user's .npmrc. 
You can view/revoke  active tokens for your user account by going to the settings page https://www.npmjs.com/settings/tokens

**Comment by soldair** [comment]
i expect you have one active token for every publish :fearful:

**Comment by ghost** [comment]
OK, I'll delete all tokens, create a new one, and use that one exclusively from here on out.
**UPDATE**: Fixed in Microsoft/types-publisher#157
